pkg/controller/azure/compute/aks.go

```diff
@@ -0,0 +1,298 @@
+	kubeclient kubernetes.Interface
+	recorder   record.EventRecorder
+
+	connect func(*azurecomputev1alpha1.AKSCluster) (aks.Client, error)
+	create  func(*azurecomputev1alpha1.AKSCluster, aks.Client) (reconcile.Result, error)
+	delete  func(*azurecomputev1alpha1.AKSCluster, aks.Client) (reconcile.Result, error)
+	sync    func(*azurecomputev1alpha1.AKSCluster, aks.Client) (reconcile.Result, error)
+	secret  func(*azurecomputev1alpha1.AKSCluster, aks.Client) (*corev1.Secret, error)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	r := &Reconciler{
+		Client:     mgr.GetClient(),
+		scheme:     mgr.GetScheme(),
+		kubeclient: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		recorder:   mgr.GetRecorder(controllerName),
+	}
+	r.connect = r._connect
+	r.create = r._create
+	r.sync = r._sync
+	r.delete = r._delete
+	r.secret = r._secret
+	return r
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Provider
+	err = c.Watch(&source.Kind{Type: &azurecomputev1alpha1.AKSCluster{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the cluster for a Provider object and makes changes based on the state read
+// and what is in the Provider.Spec
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("reconciling %s: %v", azurecomputev1alpha1.AKSClusterKindAPIVersion, request)
+	// Fetch the Provider instance
+	instance := &azurecomputev1alpha1.AKSCluster{}
+	err := r.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Create GKE Client
+	gkeClient, err := r.connect(instance)
+	if err != nil {
+		return r.fail(instance, errorClusterClient, err.Error())
+	}
+
+	// Check for deletion
+	if instance.DeletionTimestamp != nil {
+		return r.delete(instance, gkeClient)
+	}
+
+	// Add finalizer
+	util.AddFinalizer(&instance.ObjectMeta, finalizer)
+
+	// Create cluster instance
+	if instance.Status.ClusterName == "" {
+		return r.create(instance, gkeClient)
+	}
+
+	// Sync cluster instance status with cluster status
+	return r.sync(instance, gkeClient)
+}
+
+// fail - helper function to set fail condition with reason and message
+func (r *Reconciler) fail(instance *azurecomputev1alpha1.AKSCluster, reason, msg string) (reconcile.Result, error) {
+	instance.Status.UnsetAllConditions()
```

Comment:
i've started logging the failure reason/message in addition to setting it on the instance, because when user share with us their logs then we'll see the failure right there in the logs (as opposed to also having to get them to run a `kubectl get` command)

Reply:
This "failure" is a normal and expected processing flow (i.e it is not a process failure), and this failure message is saved into the status condition. 
Side note: doesn't user need to run `kubectl logs` :)

Reply:
yes, and that's the first thing most users do even unprompted.  it's nice to be able to go to just a single consistent command of `kubectl logs <crossplane-pod>`, instead of having to figure out which type or which object to run the `kubectl get` on.  I think in general having more logging wherever we can is not a bad thing to do now that we have users running it. it doesn't have to be done for this `fail` function but it's a general comment that we could use more logging.  See https://crossplane.slack.com/archives/CEG3T90A1/p1545059932056300?thread_ts=1545059800.054900&cid=CEG3T90A1 as an example

Reply:
@jbw976 I see your point. From the same slack example:
```
Conditions:
    - LastTransitionTime: 2018-12-17T14:12:08Z
      Message: 'googleapi: Error 400: Network "default" requires specifying a subnetwork.,
        badRequest'
      Reason: Failed to create new cluster

Reply:
yes running `kubectl get` on the correct object/type gives a more scoped view of the error than having to look through the entire logs, but we had to tell the user what command to run. they didn't know that already on their own, but they did know on their own to just look at the crossplane logs.

Logs can also give a sequence of events that lead to the failure, which is nice.

Reply:
Ideally, the user should look at the status of the object they have created.

`kubectl create -f file.yaml`  followed by  `kubectl get -f file.yaml -o yaml`.
With that said, currently, we do not propagate concrete resource status to the abstract resource claim. Which is something perhaps we need to fix?

Controller logs are important if you are debugging (or troubleshooting) a controller. The analogy could be, if my `pod` fails to run, Kubernetes cluster `controller` or `scheduler` or even `kubelet` log is probably the last place I would look. My choice of digging into it would most likely be:
1. `kubectl get pod -o yaml` or `kubectl describe pod`
2. `kubectl get events`